x/ecocredit/server/core/create_class_test.go

```diff
@@ -0,0 +1,101 @@
+package core
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/golang/mock/gomock"
+	"github.com/regen-network/regen-ledger/x/ecocredit"
+	"github.com/regen-network/regen-ledger/x/ecocredit/v1beta1"
+	"gotest.tools/v3/assert"
+	"testing"
+)
+
+func TestCreateClass_Valid(t *testing.T) {
+	t.Parallel()
+	s := setupBase(t)
+	any := gomock.Any()
+	s.paramsKeeper.EXPECT().GetParamSet(gomock.Any(), gomock.Any()).Do(func(ctx interface{}, p *ecocredit.Params) {
+		p.AllowlistEnabled = false
+		p.CreditClassFee = sdk.NewCoins(sdk.NewInt64Coin("foo", 20))
+		p.CreditTypes = []*ecocredit.CreditType{{Name: "carbon", Abbreviation: "C", Unit: "tonne", Precision: 6}}
+	}).AnyTimes()
+
+	s.bankKeeper.EXPECT().SendCoinsFromAccountToModule(any, any, any, any).Return(nil)
+	s.bankKeeper.EXPECT().BurnCoins(any, any, any).Return(nil)
+
+	res, err := s.k.CreateClass(s.ctx, &v1beta1.MsgCreateClass{
+		Admin:          s.addr.String(),
+		Issuers:        []string{s.addr.String()},
+		Metadata:       nil,
+		CreditTypeAbbrev: "C",
+	})
+	assert.NilError(t, err, "error creating class: %+w", err)
+	assert.Equal(t, res.ClassId, "C01")
+
+	ci, err := s.stateStore.ClassInfoStore().GetByName(s.ctx, res.ClassId)
+	assert.NilError(t, err)
+	assert.Equal(t, res.ClassId, ci.Name)
+	_, err = s.stateStore.ClassIssuerStore().Get(s.ctx, ci.Id, s.addr)
+	assert.NilError(t, err)
+}
+
+func TestCreateClass_Unauthorized(t *testing.T) {
+	t.Parallel()
+	s := setupBase(t)
+
+	s.paramsKeeper.EXPECT().GetParamSet(gomock.Any(), gomock.Any()).Do(func(ctx interface{}, p *ecocredit.Params) {
+		p.AllowlistEnabled = true
+		p.AllowedClassCreators = append(p.AllowedClassCreators, "foo")
+	}).AnyTimes()
+	_, err := s.k.CreateClass(s.ctx, &v1beta1.MsgCreateClass{
+		Admin:          s.addr.String(),
+		Issuers:        []string{s.addr.String()},
+		Metadata:       nil,
+		CreditTypeAbbrev: "C",
+	})
+	assert.ErrorContains(t, err, "is not allowed to create credit classes")
+
+	s.paramsKeeper.EXPECT().GetParamSet(gomock.Any(), gomock.Any()).Do(func(ctx interface{}, p *ecocredit.Params) {
+		p.AllowlistEnabled = false
+	}).AnyTimes()
+	_, err = s.k.CreateClass(s.ctx, &v1beta1.MsgCreateClass{
+		Admin:          s.addr.String(),
+		Issuers:        []string{s.addr.String()},
+		Metadata:       nil,
+		CreditTypeAbbrev: "C",
+	})
+	assert.ErrorContains(t, err, "is not allowed to create credit classes")
```

Comment:
If allowlist is not enabled, this error should not occur. It looks like the first params call is interfering. Two expects for the same function should not occur within the same unit test. It might be good to break this second part out of the unauthorized unit test and in this case the user should be authorized to create a credit class.

Reply:
looking back at this, im not even sure what i was trying to achieve with this case lol - its caught by the Valid test anyway, so i think it might be best just to axe it